clients/wallet/src/horizon.rs

```diff
@@ -16,6 +16,60 @@ pub type Ledger = u32;
 
 const POLL_INTERVAL: u64 = 5000;
 
+/// Interprets the response from Horizon into something easier to read.
+async fn interpret_response<T: DeserializeOwned>(response: reqwest::Response) -> Result<T, Error> {
+	if response.status().is_success() {
+		return response.json::<T>().await.map_err(Error::HttpFetchingError)
+	}
+
+	let error = match response.json::<serde_json::Value>().await {
+		Err(e) => Error::HttpFetchingError(e),
+		Ok(resp) => {
+			let unknown = "unknown";
+			let title = resp["title"].as_str().unwrap_or(unknown);
+			let status = u16::try_from(resp["status"].as_u64().unwrap_or(400)).unwrap_or(400);
+
+			match title {
+				"Transaction Failed" => {
+					let result_code =
+						resp["extras"]["result_codes"]["transaction"].as_str().unwrap_or(unknown);
+					let envelope_xdr = resp["extras"]["envelope_xdr"].as_str().unwrap_or(unknown);
+
+					Error::HorizonSubmissionError {
+						title: title.to_string(),
+						status,
+						reason: result_code.to_string(),
+						envelope_xdr: envelope_xdr.to_string(),
+					}
+				},
+				"Transaction Malformed" => {
+					let detail = resp["detail"].as_str().unwrap_or(unknown);
+					let envelope_xdr = resp["extras"]["envelope_xdr"].as_str().unwrap_or(unknown);
+
+					Error::HorizonSubmissionError {
+						title: title.to_string(),
+						status,
+						reason: detail.to_string(),
+						envelope_xdr: envelope_xdr.to_string(),
+					}
+				},
+				_ => {
+					let detail = resp["detail"].as_str().unwrap_or(unknown);
+
+					Error::HorizonSubmissionError {
+						title: title.to_string(),
+						status,
+						reason: detail.to_string(),
+						envelope_xdr: "".to_string(),
```

Comment:
Maybe we should use an `Option<>` type for this if it can be empty?

Reply:
No wait, actually this is important because the items of the error object will be used for resubmission eventually. So let's also make this an `Option<String>` for the error if possible.

Reply:
done